app/models/reject_non_governmental_email_addresses_validator.rb

```diff
@@ -0,0 +1,25 @@
+class RejectNonGovernmentalEmailAddressesValidator < ActiveModel::EachValidator
+  NON_GOVERNMENTAL_EMAIL_DOMAIN_KEYWORDS = %w[
+    aol btinternet gmail hotmail outlook yahoo
+  ].freeze
+
+  MESSAGE = "is not a workplace-provided address".freeze
+
+  def validate_each(record, attribute, value)
+    return if value.blank?
+
+    domain_part = value.split("@").last
```

Comment:
Minor: I have a feeling valid email addresses can technically have an "@" character in the username part if that part is quoted. It might be worth checking!

Reply:
Thanks, yeah, that sounds familiar... I always forget how weird email addresses can be.

But! Luckily Signon is already fairly strict with email address format, so another validation would fail if it didn't have exactly 1 `@`. And I _think_ this method is careful enough not to choke on an input that's only going to be caught by another validation rule.

(That format is configured here https://github.com/alphagov/signon/blob/main/config/initializers/devise.rb#L164 as `/\A[^@\s]+@([a-z0-9-]+\.)+[a-z0-9]+\z/`. And then that's used as you'd expect, in a Devise-provided validation, and also recently duplicated https://github.com/alphagov/signon/blob/main/app/models/batch_invitation_user.rb#L4 :smile:)